driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2285,195 @@ public void run() {
+
+            @Override
+            public void deliver(List<NodeRefreshRequest> events) {
+                Map<Host, HostEvent> hosts = new HashMap<Host, HostEvent>();
+                // only keep the last event for each host
+                for (NodeRefreshRequest req : events) {
+                    hosts.put(req.host, req.eventType);
+                }
+                for (final Entry<Host, HostEvent> entry : hosts.entrySet()) {
+                    final Host host = entry.getKey();
+                    HostEvent eventType = entry.getValue();
+                    // Make sure we  call controlConnection.refreshNodeInfo(host)
+                    // so that we have up-to-date infos on that host before adding it (so we typically
+                    // catch that an upgraded node uses a new cassandra version).
+                    switch (eventType) {
+                        case UP:
+                            // Cassandra tends to send notifications for new/up nodes a bit early (it is triggered once
+                            // gossip is up, but that is before the client-side server is up), so we add a delay
+                            // (otherwise the connection will likely fail and have to be retry which is wasteful). This
+                            // probably should be fixed C* side, after which we'll be able to remove this.
+                            scheduledTasksExecutor.schedule(new ExceptionCatchingRunnable() {
+                                @Override
+                                public void runMayThrow() throws InterruptedException, ExecutionException {
+                                    if (controlConnection.refreshNodeInfo(host)) {
+                                        triggerOnUp(host);
+                                    } else {
+                                        logger.debug("Not enough info for {}, ignoring host", host);
+                                    }
+                                }
+                            }, NEW_NODE_DELAY_SECONDS, TimeUnit.SECONDS);
+                            break;
+                        case ADDED:
+                            // Cassandra tends to send notifications for new/up nodes a bit early (it is triggered once
+                            // gossip is up, but that is before the client-side server is up), so we add a delay
+                            // (otherwise the connection will likely fail and have to be retry which is wasteful). This
+                            // probably should be fixed C* side, after which we'll be able to remove this.
+                            scheduledTasksExecutor.schedule(new ExceptionCatchingRunnable() {
+                                @Override
+                                public void runMayThrow() throws InterruptedException, ExecutionException {
+                                    if (controlConnection.refreshNodeInfo(host)) {
+                                        triggerOnAdd(host);
```

Comment:
Previously we would call onAdd() and execute the task synchronously; I believe this is not desired because this means the (single) thread of the scheduled executor. So I changed that for the (asynchronous) method triggerOnAdd().